runtime/shared_heap.c

```diff
@@ -1398,13 +1502,17 @@ static void verify_pool(pool* a, sizeclass sz, struct mem_stats* s) {
          NOT_MARKABLE, which is of no consequence for this verification
          (namely, that there is no garbage left). */
       header_t hd = Hd_hp(p);
-      CAMLassert(hd == 0 || !Has_status_hd(hd, caml_global_heap_state.GARBAGE));
-      if (hd) {
+      CAMLassert(
+        POOL_BLOCK_FREE_HD(hd) ||
+          !Has_status_hd(hd, caml_global_heap_state.GARBAGE)
+      );
+      if (!POOL_BLOCK_FREE_HD(hd)) {
         s->live += Whsize_hd(hd);
         s->overhead += wh - Whsize_hd(hd);
         s->live_blocks++;
       } else {
-        s->free += wh;
+        s->free += wh * (Wosize_hd(hd)+1);
```

Comment:
I've deliberately only used `Wosize_hd` for the run sequence because I mixed `Whsize_hd` and `Wosize_hd` before and got horribly burnt by an off-by-one.

I've moved this around slightly and added a comment to clarify for future readers: https://github.com/ocaml/ocaml/pull/13616/commits/fa28ac027a7ff7c4c2cd96e4ee538ba59c6de193